modify error message when redirected but expecting to render a template

AEgan opened the pull request:
This change modifies the error message when using `expect(response).to have_rendered(:template)` when the response is a redirect.

The current error message is
```
expecting <"template_name"> but rendering with <[]>
```

This change would modify it to be
```
expecting <'template_name'> but was a redirect to <(redirect location)>
```

This will make the problem a bit more clear.

Because the matcher calls assert_template for this, the actual exception is being passed from assert_template to have_rendered, so when the assert_template fails it was just passing the error message along. This change checks to see if the response is a redirect after the assertion fails, and if so replaces the second half of the error message with a message that gives the actual reason for the failure.

cupakromer commented:
Thanks for improving this! :blue_heart: This will definitely help debugging controller specs.

> Because the matcher calls assert_template for this, the actual exception is being passed from assert_template to have_rendered, so when the assert_template fails it was just passing the error message along.

:astonished: Yeah, at some point we really should update our matchers to no rely on those `assert_xyz` methods.

cupakromer commented on lib/rspec/rails/matchers/have_rendered.rb:
```diff
@@ -9,18 +9,32 @@ def initialize(scope, expected, message = nil)
+            match_check
+          end
+
+          def check_redirect
```
This needs a YARD doc comment.

cupakromer commented on lib/rspec/rails/matchers/have_rendered.rb:
```diff
@@ -9,18 +9,32 @@ def initialize(scope, expected, message = nil)
+            response = @scope.response
+            return unless response.respond_to?(:redirect?) && response.redirect?
+            @redirect_is = @scope.send(:normalize_argument_to_redirection, response.location)
           end
```
When the YARD doc is added, can you include some details about this `normalize_argument_to_redirection` method and why we need to use `send`?

cupakromer commented on spec/rspec/rails/matchers/have_rendered_spec.rb:
```diff
@@ -88,6 +88,24 @@ def assert_template(*); raise "oops"; end
+        end
+        def normalize_argument_to_redirection(*)
+          "http://test.host/widgets/1"
+        end
```
Instead of `def normalize_argument_to_redirection(*)` could we update this to `def normalize_argument_to_redirection(_descriptive_name_of_what_this_is`?

cupakromer commented on spec/rspec/rails/matchers/have_rendered_spec.rb:
```diff
@@ -88,6 +88,24 @@ def assert_template(*); raise "oops"; end
+            expect(response).to send(template_expectation, "template_name")
+          end.to raise_exception("expecting <'template_name'> but was a redirect to <http://test.host/widgets/1>")
+        end
+      end
```
:heart: yay for specs!

JonRowe replied:
Normally we'd have spacing around those methods and the it too, feel like cleaning it up @cupakromer ?

cupakromer replied:
Done :smile:

JonRowe replied:
:heart:

AEgan commented:
I added a YARD doc and comments, and changed the param name in the spec. Let me know if you want anything modified or if you want me to squash the commits!

cupakromer commented:
Looks good. I think we just need a change log entry. And then merge when this is green.

Also I've opened https://github.com/rspec/rspec-rails/issues/1441 so we can make this process easier in the future.

cupakromer commented:
Thanks
